### PR TITLE
Feature/add jwt authentication

### PR DESCRIPTION
**Pull request by asonni**
Redesign folder structure, add custom error handling and add JWT authentication feature.

### CODE REVIEW

**Review by yashShelatkar — CHANGES_REQUESTED** [review]
Can you please change "" to single quotes? as per style guide

**Comment by asonni** [comment]
> Can you please change "" to single quotes? as per style guide

@yashShelatkar Done.

**Comment by rgavinc** [comment]
> Can you please change "" to single quotes? as per style guide

Sorry for the confusion @asonni and @yashShelatkar but the code but the double quotes were following our style standards. See https://github.com/zero-to-mastery/drum-root-api/blob/master/.eslintrc.js. You should be able to undo your last commit.

**Comment by asonni** [comment]
> > Can you please change "" to single quotes? as per style guide
> 
> Sorry for the confusion @asonni and @yashShelatkar but the code but the double quotes were following our style standards. See https://github.com/zero-to-mastery/drum-root-api/blob/master/.eslintrc.js. You should be able to undo your last commit.

@rgavinc Done.

**Comment by aneesh4995** [comment]
looks good i approve it, just make the changes according to style guide.

**Review by aneesh4995 — APPROVED** [review]

**Review by rgavinc — CHANGES_REQUESTED** [review]
Great work Aladdin, I appreciate your efforts to better orginaize the code. However, some changes have caused the frontend to no longer be able to function correctly. After I start up docker I get the following errors when the front-end makes requests: 
image

- [ ] Make sure that the front-end can successfully make reequests.

**Comment by asonni** [comment]
> Great work Aladdin, I appreciate your efforts to better orginaize the code. However, some changes have caused the frontend to no longer be able to function correctly. After I start up docker I get the following errors when the front-end makes requests:
> image
> 
> * [ ]  Make sure that the front-end can successfully make reequests.

Sorry about that @rgavinc. I made back the old route names to match front-end requests.

**Review by rgavinc — CHANGES_REQUESTED** [review]
@asonni 
Before I look too closely at the code I like to run it first and make sure everything is working. Now upon attempting to login I get the error: `"secretOrPrivateKey must have a value"`. 
Have you been able to successfully have your code communicate correctly with the front end? If so can you provide any information we may need to know to get this to function correctly?

- [ ] Make sure code communicates correctly with front end code and make sure to document anything we may need to know to get this new code to fuction correctly.

**Review by yashShelatkar — APPROVED** [review]

**Comment by asonni** [comment]
> @asonni
> Before I look too closely at the code I like to run it first and make sure everything is working. Now upon attempting to login I get the error: `"secretOrPrivateKey must have a value"`.
> Have you been able to successfully have your code communicate correctly with the front end? If so can you provide any information we may need to know to get this to function correctly?
> 
> * [ ]  Make sure code communicates correctly with front end code and make sure to document anything we may need to know to get this new code to fuction correctly.

@rgavinc 
You should create a new file under the name `config.env` inside the config folder and copy what's inside `config.example.env` file and then change the values as the following comments in it. I've updated README.md file - Thanks.

**Review by rgavinc — CHANGES_REQUESTED** [review]
I'm still unable to login whether or not I use docker or npm to start the back-end service. Let's chat on discord about how we can cover this task.

File: config/config.example.env [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Default development
```
Comment (rgavinc) [inline config/config.example.env]: - [ ] Can we use `.env` instead and set all the default values so people do not need to configure this themselves?

**Comment by mrpacino** [comment]
> I'm still unable to login whether or not I use docker or npm to start the back-end service. Let's chat on discord about how we can cover this task.

ah glad i wasn't the only one that had this problem!

File: middleware/auth.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  let token;
+  const { authorization } = req.headers;
+
+  if (authorization && authorization.startsWith("Bearer")) {
+    // Set token from Bearer token in header
+    token = authorization.split(" ")[1];
+    // Set token from cookie
+  }
+  // else if (req.cookies.token) {
+  //   token = req.cookies.token;
```
Comment (rgavinc) [inline middleware/auth.js]: - [ ] Either remove this code or add a "TODO:" to help us understand what we want to do here in the future.

File: middleware/error.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  if (err.isOperational) {
+    res.status(err.statusCode).json({
+      success: false,
+      message: err.message
+    });
+    // Programming or other unknown error: don't leak error details
+  } else {
+    // 1) Log error
+    // eslint-disable-next-line no-console
+    console.error("ERROR 💥", err);
```
Comment (rgavinc) [inline middleware/error.js:21]: I like the explosion :D

File: server.js [hunk 4]
```diff
@@ -1,28 +1,46 @@
...
+app.all("*", (req, res, next) => {
+  next(new ErrorResponse(`Can't find ${req.originalUrl} on this server!`, 404));
 });
 
-app.post("/signin", handleSignin);
-app.post("/signout", handleSignout);
-app.get("/drumlayout", requireAuth, getLayout);
+app.use(errorHandler);
 
-const PORT = process.env.PORT || 3000;
```
Comment (rgavinc) [inline server.js:25]: - [ ] Let's switch this to just plan old `process.env.PORT` so that this will work in different future environments (dev vs prod)

**Review by rgavinc — CHANGES_REQUESTED** [review]
This is looking really good! I just have two more small requests.

File: server.js (hunk 4 above)
Comment (asonni) [reply server.js:25]: Done

File: middleware/auth.js (hunk 2 above)
Comment (asonni) [reply middleware/auth.js]: Done

**Review by rgavinc — APPROVED** [review]
Good work!

**Comment by mrpacino** [comment]
good stuff! all working for me, seeing as the others have signed off. off to merge it goes!

**Pull request merged**